Reduce UI thread consumption

allisonchou opened the pull request:
### Summary of the changes
- Today, the UI thread is used unnecessarily to run code in some areas. These areas primarily use the UI thread not because they need the UI thread specifically, but because they require running on a single thread. As a solution, we now implement a custom "fake" single-threaded foreground thread, which the previously mentioned code can run on. (Creds to Taylor for the genius idea!)
- There are some areas that require the actual UI thread, so we can't get rid of the dependency completely.
 -  (Fixed) <s>The tests haven't been properly fixed up yet, but I just wanted to check that this general approach was going OK before going through and fixing all of them.</s>
 - (Fixed) <s>I think there's some sort of issue with discovering components defined within a project. Component discovery seems to only work part of the time? I feel like it may due to some sort of race condition. If anyone has any ideas please let me know!</s>

Fixes: 
https://github.com/dotnet/aspnetcore/issues/32975

allisonchou commented on src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Microbenchmarks/ProjectSystem/ProjectSnapshotManagerBenchmarkBase.cs:
```diff
@@ -93,9 +93,11 @@ private static IReadOnlyList<TagHelperDescriptor> ReadTagHelpers(string filePath
         private class TestForegroundDispatcher : ForegroundDispatcher
         {
-            public override bool IsForegroundThread => true;
+            public override bool IsSpecializedForegroundThread => true;
```
I renamed this from `IsForegroundThread` -> `IsSpecializedForegroundThread` to better reflect that we're referring to a specialized thread here as opposed to the UI thread.

allisonchou commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer.Common/BackgroundDocumentGenerator.cs:
```diff
@@ -136,7 +138,7 @@ private void OnBackgroundCapturedWorkload()
         internal void Enqueue(DocumentSnapshot document)
         {
-            _foregroundDispatcher.AssertForegroundThread();
+            _foregroundDispatcher.AssertSpecializedForegroundOrUIThread();
```
Wherever `AssertSpecializedForegroundOrUIThread` is called indicates code that may be executed by either the specialized foreground thread or the UI thread depending on the situation.

I'm not super confident on these asserts (I basically just looked at the callsites and made educated guesses, and made sure no exceptions were thrown when running), so if any of these asserts look off, e.g. supposed to run exclusively on one thread, let me know!

allisonchou commented on src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test.Common/LanguageServerTestBase.cs:
```diff
@@ -35,17 +35,19 @@ private class SingleThreadedForegroundDispatcher : ForegroundDispatcher
             public SingleThreadedForegroundDispatcher()
             {
-                ForegroundScheduler = SynchronizationContext.Current == null ? new ThrowingTaskScheduler() : TaskScheduler.FromCurrentSynchronizationContext();
+                SpecializedForegroundScheduler = SynchronizationContext.Current == null ? new ThrowingTaskScheduler() : TaskScheduler.FromCurrentSynchronizationContext();
```
All files including and below this file can be ignored, as they're all test files and I don't have them set up to pass yet 😃

allisonchou commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs:
```diff
@@ -159,8 +170,19 @@ private ProjectSnapshotManagerBase GetProjectManager()
+            {
+                using (await _updateLock.EnterAsync().ConfigureAwait(false))
+                {
+                    await _foregroundDispatcher.SpecializedForegroundScheduler;
```
The primary change in the PR is here (going from using the UI thread to this specialized thread)

allisonchou commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ForegroundDispatcher.cs:
```diff
@@ -20,13 +26,24 @@ public virtual void AssertForegroundThread([CallerMemberName] string caller = nu
             if (!IsForegroundThread)
```
I eventually want to rename `IsForegroundThread` to `IsSpecializedForegroundThread` and `AssertForegroundThread` to `AssertSpecializedForegroundThread` to clearly indicate that the referred-to foreground thread is distinct from the UI thread. However, I tried doing this and the diff was huge, so I plan to do it in a follow-up PR.

allisonchou replied:
Will do in a follow-up!

allisonchou commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs:
```diff
@@ -155,22 +158,25 @@ public override void DocumentAdded(HostProject hostProject, HostDocument documen
+            _foregroundDispatcher.AssertSpecializedForegroundOrUIThread();
 
-            if (_projects.TryGetValue(hostProject.FilePath, out var entry))
+            lock (_lock)
```
All the code in this file is the same. However, in order to account for the possibility that the methods can be either called from the specialized foreground thread or the UI thread, I put any code that writes to a field variable into a lock to prevent concurrent modification.

allisonchou commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs:
```diff
@@ -159,8 +170,19 @@ private ProjectSnapshotManagerBase GetProjectManager()
+            {
+                using (await _updateLock.EnterAsync().ConfigureAwait(false))
+                {
+                    await _foregroundDispatcher.ForegroundScheduler;
```
This is the primary change in the PR (going from using UI thread -> specialized foreground scheduler).

allisonchou commented on src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test.Common/LanguageServerTestBase.cs:
```diff
@@ -46,6 +46,8 @@ public SingleThreadedForegroundDispatcher()
             private Thread Thread { get; } = Thread.CurrentThread;
 
             public override bool IsForegroundThread => Thread.CurrentThread == Thread;
+
```
Anything below this point can be ignored, it's all tests and I don't have them set up to work properly currently

allisonchou commented:
@NTaylorMullen @dotnet/razor-tooling This is somewhat ready for review. I still can't figure out why exactly components defined within a project aren't being discovered some of the time. I suspect it's a race condition somewhere between the UI thread and the specialized foreground thread, but I can't solve it despite trying to put locks over everything that looks suspect. ☹️ Or maybe there's some code here that actually requires the UI thread? I'm not sure, but would definitely appreciate any ideas since I feel like I'm out of them.

The PR is a lot longer than it looks - lots of code being moved around, and resource string changes.

allisonchou commented:
Also, this area is somewhat new to me so I'm fully expecting lots of feedback. 😄 Feel free to be harsh.

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs:
```diff
@@ -159,8 +170,19 @@ private ProjectSnapshotManagerBase GetProjectManager()
-            action();
+            // The passed-in action is intended to execute solely on our specialized single-threaded
+            // foreground scheduler. However, there are times when the action may need to temporarily
+            // switch to the UI thread to perform a certain task. We want to make sure that no other
```
Hmmm, not sure I follow. Could you provide an example? Or if chatting offline is easier we can do that too 😄

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioForegroundDispatcher.cs:
```diff
@@ -2,20 +2,23 @@
+        }
 
-        public override bool IsForegroundThread => ThreadHelper.CheckAccess();
+        public override bool IsBackgroundThread => !IsForegroundThread && !_joinableTaskContext.IsOnMainThread;
```
Oh ok, this is the bit that allows the UI thread potentially.

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/VisualStudioMacEditorDocumentManagerFactory.cs:
```diff
@@ -15,16 +16,20 @@ namespace Microsoft.VisualStudio.Mac.LanguageServices.Razor
 
         [ImportingConstructor]
-        public VisualStudioMacEditorDocumentManagerFactory(ForegroundDispatcher foregroundDispatcher)
+        public VisualStudioMacEditorDocumentManagerFactory(
```
Nope you right!

NTaylorMullen reviewed (commented):
Exciting to see this come together! One thing also to validate is to use the older Razor editor and ensure things still work. It relies a little bit more on the non-LanguageServer bits then our new stuff

ryanbrandenburg commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultForegroundDispatcher.cs:
```diff
@@ -5,24 +5,25 @@
-namespace Microsoft.AspNetCore.Razor.LanguageServer.Common
+namespace Microsoft.CodeAnalysis.Razor.Workspaces
 {
     internal class DefaultForegroundDispatcher : ForegroundDispatcher
```
Yeah I agree that `Foreground` doesn't seem to capture the intent here anymore. Maybe something like `SerialDispatcher` or `SingleThreadedDispatcher`? `Serial` might be a no-go because it's a bit of an overload with the O# Serial vs Parallel concepts.

ryanbrandenburg commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ForegroundDispatcher.cs:
```diff
@@ -20,13 +26,24 @@ public virtual void AssertForegroundThread([CallerMemberName] string caller = nu
+        // run on the UI thread or the specialized foreground thread.
+        public virtual void AssertSpecializedForegroundOrUIThread([CallerMemberName] string caller = null)
+        {
+            if (IsBackgroundThread)
```
Yeah, seems like this could be renamed to align with whatever terminology we land on and remove the "Or".

ryanbrandenburg commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Documents/EditorDocumentManagerListener.cs:
```diff
@@ -61,14 +86,21 @@ public override void Initialize(ProjectSnapshotManagerBase projectManager)
+                        // GetOrCreateDocument requires the UI thread due to IVs access, so we need to temporarily switch to it.
+                        await _joinableTaskFactory.SwitchToMainThreadAsync();
                         var document = _documentManager.GetOrCreateDocument(key, _onChangedOnDisk, _onChangedInEditor, _onOpened, _onClosed);
+                        await _foregroundDispatcher.ForegroundScheduler;
```
Yeah, I'd prefer something more explicit than awaiting the Scheduler itself, even if all it does on the inside is await the Scheduler.

ryanbrandenburg commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Documents/VisualStudioEditorDocumentManager.cs:
```diff
@@ -63,6 +61,8 @@ internal class VisualStudioEditorDocumentManager : EditorDocumentManagerBase
 
         protected override ITextBuffer GetTextBufferForOpenDocument(string filePath)
         {
+            JoinableTaskContext.AssertUIThread();
```
Maybe we should add comments above each call to AssertUIThread with the reason/call that it's needed for? I would GUESS it's the calls to `_runningDocumentTable` in this instance, but I'm not sure.

ryanbrandenburg commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs:
```diff
@@ -159,8 +170,19 @@ private ProjectSnapshotManagerBase GetProjectManager()
+            // protect the below area with a lock.
+            using (JoinableCollection.Join())
+            {
+                using (await _updateLock.EnterAsync().ConfigureAwait(false))
```
Doesn't this lock only protect us from losing the SpecialForeground thread within `RazorProjectHostBase`?

ryanbrandenburg reviewed (commented):
I got lost in a couple places, but I suspect that I just don't understand rather than there's anything wrong with the design.

allisonchou commented:
I think I've addressed most of the feedback, thanks so much Taylor and Ryan!

A couple of notes about the existing state of the PR:
- The issue with unrecognized components has gone away, but now we're getting stale diagnostics on components. I feel like this is probably still due to some sort of race condition between the UI thread and specialized foreground thread. Potentially it's the issue I detailed in [this comment](https://github.com/dotnet/aspnetcore-tooling/pull/3808/commits/d799de0f4f96e657f83616551c7afbe77d8ff723#r654882643)? (Also pasted below)
> Basically, there are some places within UpdateAsync where we need to switch from our specialized single-threaded thread to the UI thread (due to IVs access). I'm wondering if this could be problematic since we temporarily "lose" the specialized single-thread while we're on the UI thread, and the specialized single-threaded thread can potentially in the meantime be picked up by some other code that's also waiting for it.

Issue screenshot:
![image](https://user-images.githubusercontent.com/16968319/122665659-5ac5ac80-d15d-11eb-9187-dbf0639ad32c.png)
- Tests are still broken (on purpose) - there's quite a few that need to be fixed up, so just want to get the sign off on the implementation before going through and fixing them
- Planning on doing all the rename work in a follow-up PR

allisonchou commented:
> I got lost in a couple places, but I suspect that I just don't understand rather than there's anything wrong with the design.

Lmk if there's anything I can help clarify. I don't know this area well either, kind of just paddling along and learning as I go. 😄

NTaylorMullen commented:
> I'm not sure, but would definitely appreciate any ideas since I feel like I'm out of them.

Want to schedule some time for us both to dig through the details? Could be fun to pair debug it 😄

NTaylorMullen commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ForegroundDispatcher.cs:
```diff
@@ -11,6 +11,12 @@ internal abstract class ForegroundDispatcher
+        public abstract bool IsBackgroundThread { get; }
+
+        /// <summary>
+        /// Single-threaded scheduler intended to be a "fake" UI thread for code that requires
```
> 3\. Rename `BackgroundScheduler` to `ReaderScheduler`

While I wish this was 100% true one interesting implication is that currently the ProjectSnapshtomanager requires you to ber on the "Writer" in order to read any data from it; I wonder if we'd be able to change that restriction. Anyhow, problem for another day 😄

NTaylorMullen commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/WorkspaceProjectStateChangeDetector.cs:
```diff
@@ -10,6 +10,7 @@
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Razor.Language.Components;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.VisualStudio.Threading;
```
So this is the magic that's enabling you to `await` a `TaskScheduler`. Typically that wouldn't be possible, and in our language server we don't depend on Visual Studio at all which would make depending on the `ForegroundDispatcher`'s `SwitchToForeground` method less seamless 😄. I know I suggested to have a `SwitchToForeground` method but now I'm thinking maybe a `RunOnForegroundAsync<TParam>(Action<TParam> action)` may be a better route to take so we fully understand what we're running on the foreground (since it'll be encapsulated in the `Task`) and what we're not. That way you can utilize `Task.Factory.StartNew` under the covers too.

ryanbrandenburg commented:
> > I got lost in a couple places, but I suspect that I just don't understand rather than there's anything wrong with the design.
> 
> Lmk if there's anything I can help clarify. I don't know this area well either, kind of just paddling along and learning as I go. 😄

I think I was mostly confused about the lock, which you've changed anyway.

allisonchou commented:
I think this PR should be ready for review again. I switched more portions of the code off the UI thread, and now we only use it in two places. Also, the tests should be fixed up now.

ryanbrandenburg commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Documents/EditorDocumentManagerBase.cs:
```diff
@@ -197,22 +206,25 @@ public sealed override void RemoveDocument(EditorDocument document)
 
-            var key = new DocumentKey(document.ProjectFilePath, document.DocumentFilePath);
-            if (_documentsByFilePath.TryGetValue(document.DocumentFilePath, out var documents))
+            lock (_lock)
```
Good catch.

Maybe not for this PR, but would this be better if we did `ConcurrentDictionary` instead of `Dictionary`? I'm thinking it might not work because this call probably needs the TryGetValue and Remove calls on `_documentsByFilePath` to be atomic, but it'd be a shame to use locks if `ConcurrentDictionary` would do (to avoid mistakes like the one you're fixing here).

allisonchou replied:
Good idea! Yeah, I would assume those operations would have to be atomic, but if we could avoid using locks here somehow that would be ideal.

ryanbrandenburg reviewed (approved):
Looks good, couple nitpicks.

Does it feel any snappier when you run the experimental instance (with debugger detached)?

allisonchou commented:
Thanks for the review Ryan!

> Does it feel any snappier when you run the experimental instance (with debugger detached)?

Honestly, both before & after feel pretty snappy to me in terms of responsiveness, so I didn't notice a huge difference, even with a larger solution. 😅 Although I pair-debugged with Taylor on Wednesday and I _think_ he mentioned noticing a difference, so maybe it's just me?

Side note, I only now just remembered Taylor's reminder to test with the old Editor, and things are def still broken there. 🤦‍♀️ looking into fixing things up now.

allisonchou commented:
Ahh, I've run into an issue when trying to fix things up in the old editor. 😦 

Essentially, WebTools calls into one of our Razor methods, `TryGetDocumentTracker`, [here](http://index/?rightProject=Microsoft.WebTools.Languages.Razor.Core&file=RazorCodeGenerator.cs&line=155). They call us from the UI thread, which is problematic since there's some code nested in `TryGetDocumentTracker` that should be run exclusively on the specialized single thread. One option would be to change the method signature of `TryGetDocumentTracker` to be async so we can switch to the specialized single thread . However, since WebTools calls into the method, I don't believe we can change the method signature without a coordinated dual insertion.

Some potential solutions:
1) Dual insertion with WebTools where we change `TryGetDocumentTracker`'s method signature to async. This option seems like it may still be problematic because then we'd have to change other areas in WebTools (due to the async change), and I'm not sure of the ramifications stemming from that.
2) Make some of our specialized single thread assert checks more "lax" temporarily until we get rid of the old editor completely. Unfortunately, this means that the old editor could be accessing resources both via the specialized single thread and the UI thread. Also, by making the checks more "lax", we won't be able to assert that we're doing the right things in the LSP scenario.

Thoughts or any additional ideas @NTaylorMullen @ryanbrandenburg? Personally, I'm leaning towards the second option since we plan to eventually phase out the new editor anyway. But it could potentially make the old editor even buggier, and getting rid of some of the assert checks makes me worried since the checks affect the LSP scenario as well.

ryanbrandenburg commented:
Is there maybe potential solution 3: Put the new behavior in a new method `TryGetDocumentTrackerAsync` but leave the old behavior for `TryGetDocumentTracker`, then switch WebTools to the new method at our leisure (if possible)?

The big problem with that would be if the intertwined-ness goes too deep and maintaining both versions would be a nightmare, do you think that's likely?

I agree that laxing the asserts seems dangerous, but maybe the amount of stuff you'd have to touch in a Dual-insertion is worse? If it's feasible I think option 3 would be better, but I wouldn't be surprised if it's not feasible (or if I don't understand the 
 area).

allisonchou commented:
I really like the idea! However, I didn't expand much in my original message, and I think your comment here is apt:

> The big problem with that would be if the intertwined-ness goes too deep and maintaining both versions would be a nightmare, do you think that's likely?

I think this could be the main blocker for adopting this approach :( So the failing assert is all the way in `DefaultProjectSnapshotManager` (10-15 calls in from `TryGetDocumentTracker`). With this approach, I think we'd have to create an entirely new `ProjectSnapshotManager`, one for the new editor and one for the old editor, which would basically be identical other than the assert statements. Also, we'd have to somehow make one snapshot manager only activate in the LSP scenario, and the other only activate in the legacy scenario. (Maybe there's an easy way to do this, but looking at the code it seems a bit convoluted.)

ryanbrandenburg commented:
> 10-15 calls in from TryGetDocumentTracker
Oh yeah, that's gonna be a huge mess, disregard options 3.

In that case I'm torn. Though I feel option 1 is the better answer technically I can see it being a lot of work, and the fact that it's not your first choice makes me think it's even more work (and even more danger/breakage potential) than I'm imagining. I'll be interested to hear @NTaylorMullen's thoughts

NTaylorMullen commented:
Oh good old document tracker lmao. Sooo this is going to sound a little batshit, and it is to an extent, but what if we were to synchronously switch to our own fake UI thread? Given how infrequently we use it My bet is that 99% of the time the swap will be synchronous and the other 1% of the time there's no chance for it to block in a way that would cause a deadlock.

allisonchou commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Resources.resx:
```diff
@@ -129,4 +129,7 @@
   <data name="ForegroundDispatcher_NoMethodNamePlaceholder" xml:space="preserve">
     <value>The method</value>
   </data>
+  <data name="JoinableTaskContextExtensions_AssertUIThread" xml:space="preserve">
```
fwiw, I don't believe Roslyn localizes error messages (or at least, I haven't been doing that, lol).

allisonchou replied:
Actually I just checked and I was totally wrong, Roslyn _does_ localize certain error messages, at least the yellow bar ones (e.g. "so-and-so code fix has been disabled"). However, specific details about what went wrong (i.e. stack trace and specific error message) don't seem to get localized.

NTaylorMullen replied:
In our tooling stack we don't currently present any yellow bars directly so I'd actually say we should avoid localizing these if possible because we'll be the ones needing to decipher the error messages in reported issues 😄

allisonchou replied:
Sounds good! I removed the added ones in my latest push, but the left the existing ones for now.

NTaylorMullen commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ForegroundDispatcher.cs:
```diff
@@ -11,10 +12,18 @@ internal abstract class ForegroundDispatcher
         public abstract TaskScheduler BackgroundScheduler { get; }
 
+        public Task RunOnForegroundAsync(Action action, CancellationToken cancellationToken)
+            => RunOnForegroundAsync(new Func<Task>(() => { action(); return Task.CompletedTask; }), cancellationToken);
```
I would just call Task.Factory.StartNew separately instead of trying to utilize a Func. The Task.Factory.STartNew method has an overload that takes an Action is why

NTaylorMullen commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/JoinableTaskContextExtensions.cs:
```diff
@@ -0,0 +1,46 @@
+            var result = action();
+
+            // Return to original thread
+            await originalScheduler;
```
Hmmm, is the intent behind this method to be consistent with our RunOnForegroundAsync? While I think this is actually a nice idea given it allows us to control exactly what's run on the UI thread I'm not 100% convinced it's needed only because VS paradigms have gotten us used to the SwithcTOMainThreadAsync paradigm. Although I'm up for being convinced otherwise 😄

allisonchou replied:
Gotcha. Yep, the original intent was to be consistent with RunOnForegroundAsync, but I don't have a strong preference either way. I'll change it back to how it was

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -94,68 +106,90 @@ internal void TriggerSmartIndent(ITextView textView)
             {
-                return;
-            }
+                await _foregroundDispatcher.RunOnForegroundAsync(() =>
```
Ah IIRC this can run on the UI thread like it did before since I don't believe it interacts with our ProjectSnapshotManager

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -94,68 +106,90 @@ internal void TriggerSmartIndent(ITextView textView)
-                var caret = textView.Caret.Position.BufferPosition;
-                var textViewBuffer = textView.TextBuffer;
-                var indent = CalculateIndent(textViewBuffer, context.ChangePosition);
+                await _foregroundDispatcher.RunOnForegroundAsync(() =>
```
Ditto

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultBraceSmartIndenterFactory.cs:
```diff
@@ -45,9 +54,9 @@ public override BraceSmartIndenter Create(VisualStudioDocumentTracker documentTr
+            _joinableTaskContext.AssertUIThread();
 
-            var braceSmartIndenter = new BraceSmartIndenter(_dispatcher, documentTracker, _codeDocumentProvider, _editorOperationsFactory);
+            var braceSmartIndenter = new BraceSmartIndenter(_foregroundDispatcher, _joinableTaskContext, documentTracker, _codeDocumentProvider, _editorOperationsFactory);
```
Probably doesn't need the foreground dispatcher at all actually 😄

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultBraceSmartIndenterFactoryFactory.cs:
```diff
@@ -15,43 +16,51 @@ namespace Microsoft.VisualStudio.Editor.Razor
             }
 
-            return new DefaultBraceSmartIndenterFactory(_foregroundDispatcher, _codeDocumentProvider, _editorOperationsFactory);
+            return new DefaultBraceSmartIndenterFactory(_foregroundDispatcher, _joinableTaskContext, _codeDocumentProvider, _editorOperationsFactory);
```
ditto

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultRazorDocumentManager.cs:
```diff
@@ -20,31 +20,26 @@ internal class DefaultRazorDocumentManager : RazorDocumentManager
         [ImportingConstructor]
         public DefaultRazorDocumentManager(
-            ForegroundDispatcher dispatcher,
+            ForegroundDispatcher foregroundDispatcher,
```
Pretty sure this doesn't need the dispatcher anymore. I'd imagine this would expect everything to be UI Thread related. One of the key reasons is that a lot of the older stack didn't know about the ProjectSNapshotManager at all but utilized the ForegroundDispatcher as a mechanism to assert "is ui thread" which was a bit backwards 😄. We made big mistakes in the past

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioDocumentTracker.cs:
```diff
@@ -32,7 +37,8 @@ internal class DefaultVisualStudioDocumentTracker : VisualStudioDocumentTracker
 
         public DefaultVisualStudioDocumentTracker(
-            ForegroundDispatcher dispatcher,
+            ForegroundDispatcher foregroundDispatcher,
```
Ugh I hate the reason why we need this here. Lol past us really sucks

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioRazorParser.cs:
```diff
@@ -366,9 +376,9 @@ internal void OnIdle(object state)
 
         // Internal for testing
-        internal void ReparseOnForeground(object state)
+        internal void ReparseOnUIThread(object state)
```
👏

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioRazorParser.cs:
```diff
@@ -366,9 +376,9 @@ internal void OnIdle(object state)
+        internal void ReparseOnUIThread(object state)
         {
-            _dispatcher.AssertForegroundThread();
+           _joinableTaskContext.AssertUIThread();
```
Judging by the call sites in this file it looks like you're dispatching to this method via the _dispatcher. For what it's worth I actually don't think this class needs the _dispatcher at all. Might make this easier.

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Documents/EditorDocumentManagerListener.cs:
```diff
@@ -61,56 +78,126 @@ public override void Initialize(ProjectSnapshotManagerBase projectManager)
-                    }
-
-                case ProjectChangeKind.DocumentRemoved:
+                await _foregroundDispatcher.RunOnForegroundAsync(async () =>
```
You should already be on the fake foreground thread. ProjectManager_Changed events should alwyas be dispatched on it

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Documents/VisualStudioEditorDocumentManager.cs:
```diff
@@ -26,55 +30,59 @@ internal class VisualStudioEditorDocumentManager : EditorDocumentManagerBase
                 throw new ArgumentNullException(nameof(filePath));
             }
 
+            ForegroundDispatcher.AssertForegroundThread();
```
Hmm, any reason for this path to require that it's on the foreground thread to start? I imagine this entire call path may want to be on the UI thread since it's interacting with the running doc table.

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Documents/EditorDocumentManager.cs:
```diff
@@ -2,14 +2,15 @@
     internal abstract class EditorDocumentManager : IWorkspaceService
     {
-        public abstract EditorDocument GetOrCreateDocument(
+        public abstract Task<EditorDocument> GetOrCreateDocumentAsync(
```
Hmm, I'm curious why this needs to be async now. I'd imagine this would require the UI thread to be called.

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Documents/RunningDocumentTableEventSink.cs:
```diff
@@ -2,51 +2,83 @@
-            if ((grfAttribs & (uint)__VSRDTATTRIB.RDTA_MkDocument) != 0)
+        private async Task OnAfterAttributeChangeExAsync(uint docCookie, uint grfAttribs, string pszMkDocumentOld, string pszMkDocumentNew)
+        {
+            await _foregroundDispatcher.RunOnForegroundAsync(async () =>
```
Might be able to get away with having this just be on the UI thread. I'd be curious to know the reason why DocumentOpenedAynsc required the fake foreground thread

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs:
```diff
@@ -157,11 +166,7 @@ private ProjectSnapshotManagerBase GetProjectManager()
-            await CommonServices.ThreadingService.SwitchToUIThread();
-            action();
-        }
+        protected async Task UpdateAsync(Action action) => await _foregroundDispatcher.RunOnForegroundAsync(action, CancellationToken.None);
```
```suggestion
        protectedTask UpdateAsync(Action action) => _foregroundDispatcher.RunOnForegroundAsync(action, CancellationToken.None);
```

NTaylorMullen commented:
Went threw the latest updates. I know it wasn't 100% ready for review quite yet but figured I'd glaze over some of the bits to potentially provide some early insights on some of the legacy editor stack given it's crazzzzzyyy 😄

allisonchou commented:
I think this should be ready for review again. I'm having trouble fixing up some tests, but other than that I think things should be in a good state for reviews. I took the approach Taylor recommended of just changing the project snapshot manager to use the fake UI thread, and then responding to the fallout from that in both the legacy + LSP editors by modifying code accordingly.

NTaylorMullen commented on src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Microbenchmarks/ProjectSystem/ProjectSnapshotManagerBenchmarkBase.cs:
```diff
@@ -96,8 +96,6 @@ private class TestForegroundDispatcher : ForegroundDispatcher
 
             public override TaskScheduler ForegroundScheduler => TaskScheduler.Default;
-
-            public override TaskScheduler BackgroundScheduler => TaskScheduler.Default;
```
Nah I love that you ridded us of it 👏

NTaylorMullen commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultHostDocumentFactory.cs:
```diff
@@ -64,7 +54,7 @@ void GeneratedDocumentContainer_Changed(object sender, TextChangeEventArgs args)
                     CancellationToken.None,
                     TaskCreationOptions.None,
-                    _foregroundDispatcher.BackgroundScheduler);
+                    TaskScheduler.Default);
```
👍

NTaylorMullen commented on src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultForegroundDispatcher.cs:
```diff
@@ -5,24 +5,21 @@
     internal class DefaultForegroundDispatcher : ForegroundDispatcher
     {
-        public override bool IsForegroundThread => Thread.CurrentThread.ManagedThreadId == ForegroundTaskScheduler.Instance.ForegroundThreadId;
+        public override bool IsForegroundThread => System.Environment.CurrentManagedThreadId == ForegroundTaskScheduler.Instance.ForegroundThreadId;
```
I actually had no idea the `System.Environment.CurrentManagedThreadId` API existed. Is it any different than `Thread.CurrentThread.ManagedThreadID`? I lean towards using the latter only for consistencies sake but don't know enough

allisonchou replied:
I have no idea, tbh I changed it because it was actually an analyzer suggestion. 😅 looking at the docs I can't see any reason why `Thread.CurrentThread.ManagedThreadId` shouldn't be used though, so I can change it back.

NTaylorMullen replied:
Ya I'd change it back just for consistency with the past / certainty it does what we think 😄

allisonchou replied:
Sounds good! I also filed https://github.com/dotnet/roslyn-analyzers/issues/5197 because I have no idea why the diagnostic was being offered, and the diagnostic info page gives an error

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioRazorParser.cs:
```diff
@@ -147,19 +149,19 @@ public override void QueueReparse()
             else
             {
-                _ = Task.Factory.StartNew(ReparseOnForeground, null, CancellationToken.None, TaskCreationOptions.None, _dispatcher.ForegroundScheduler);
+                _ = Task.Factory.StartNew(ReparseOnUIThread, CancellationToken.None, TaskCreationOptions.None, TaskScheduler.FromCurrentSynchronizationContext());
```
It does not sadly. It'd only return it if this was being called onthe UI thread, that being said, given this isn't breaking it's highly likely this code is never called? 

You'll want to change this to swap to the main thread and then call by awaiting the SwitchToMainThreadAsync bit in a fire& forget async method

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioRazorParserFactoryFactory.cs:
```diff
@@ -6,24 +6,25 @@
 
         [ImportingConstructor]
-        public DefaultVisualStudioRazorParserFactoryFactory(ForegroundDispatcher foregroundDispatcher)
+        public DefaultVisualStudioRazorParserFactoryFactory(ForegroundDispatcher foregroundDispatcher, JoinableTaskContext joinableTaskContext)
```
```suggestion
        public DefaultVisualStudioRazorParserFactoryFactory(JoinableTaskContext joinableTaskContext)
```

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Documents/EditorDocument.cs:
```diff
@@ -75,7 +95,8 @@ internal sealed class EditorDocument : IDisposable
             {
-                _fileTracker.StartListening();
+                _ = _foregroundDispatcher.RunOnForegroundAsync(
+                    () => _fileTracker.StartListening(), CancellationToken.None);
```
Good thing to have a delimma on 😄. So I wonder how much the fileTracker bits need the UI thread to begin with. I know they utilize VS components and if those components don't require the UI thread then maybe this is fine. If they do then we may need to make the file tracker UI thread affinitized

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Documents/VisualStudioFileChangeTracker.cs:
```diff
@@ -78,7 +84,7 @@ public override void StartListening()
             }
 
-            _fileChangeAdviseTask = _joinableTaskFactory.RunAsync(async () =>
+            _fileChangeAdviseTask = _joinableTaskContext.Factory.RunAsync(async () =>
```
Oh wow, the `_fileChangeService` is free-threaded and doesn't require the UI thread?

allisonchou replied:
It seems like it! I didn't run into any trouble with it not running on the UI thread

NTaylorMullen commented on src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs:
```diff
@@ -157,11 +166,7 @@ private ProjectSnapshotManagerBase GetProjectManager()
-            await CommonServices.ThreadingService.SwitchToUIThread();
-            action();
-        }
+        protected Task UpdateAsync(Action action) => _foregroundDispatcher.RunOnForegroundAsync(action, CancellationToken.None);
```
The magic! That being said, should this take a cancellation token now? We didn't do it before which was probably just an oversight?

NTaylorMullen commented on src/Razor/test/Microsoft.VisualStudio.Editor.Razor.Test/RazorTextViewConnectionListenerTest.cs:
```diff
@@ -15,12 +15,13 @@ public class RazorTextViewConnectionListenerTest : ForegroundDispatcherTestBase
         public void SubjectBuffersConnected_CallsRazorDocumentManager_OnTextViewOpened()
         {
             // Arrange
+
```
:fire:

NTaylorMullen commented on src/Razor/test/Microsoft.VisualStudio.Editor.Razor.Test/Shared/ForegroundDispatcherWorkspaceTestBase.cs:
```diff
@@ -2,11 +2,14 @@
     {
         internal ForegroundDispatcher Dispatcher { get; } = new SingleThreadedForegroundDispatcher();
+
+        internal JoinableTaskContext JoinableTaskContext { get; } = new JoinableTaskContext();
```
I forget, does VS complain about initiating joinable task contexts like this? If not then no complaints. FOr some reason I'm recalling some unique way to instantiate joinable task contexts at test time

allisonchou replied:
You were right, it wasn't working properly. I changed it so now things should be good

NTaylorMullen reviewed (approved):
Super solid @allisonchou! So excited to see this come in.

One last thing we may need to follow up with is how this work impacts VS4Mac. I presume everything will be fine but probably worth checking once this is in. Can be an after-the-fact validation

allisonchou commented on src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/ProjectSystem/WorkspaceProjectStateChangeDetectorTest.cs:
```diff
@@ -107,14 +110,17 @@ public async Task WorkspaceChanged_SolutionEvents_EnqueuesUpdatesForProjectsInSo
             // Act
-            detector.Workspace_WorkspaceChanged(Workspace, e);
-            await Dispatcher.ForegroundScheduler; // Wait for event to finish processing on thread
+            await Dispatcher.RunOnForegroundAsync(() => detector.Workspace_WorkspaceChanged(Workspace, e), CancellationToken.None);
+            await Dispatcher.ForegroundScheduler;
```
This is super weird and I'm not sure if it's the best approach here, but it works (at least from the couple of times I ran it) - otherwise the assert below will execute before the Workspace_WorkspaceChanged has finished.

allisonchou replied:
Edit: jk, it doesn't work - exploring alternatives now
